OSDOCS#13342: OLM v1 update example outputs, CRs, and nits

michaelryanpeter opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.
Do not create or rename a top-level directory (or any subdirectory in a directory that contains a hugebook.flag file) in the repository and topic map without checking with a docs program manager first.
If a book is being created or modified, there are changes on the Customer Portal that must also be made.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s): 4.18
<!--- Specify the version or versions of OpenShift your PR applies to. -->

Issue: [OSDOCS-13342](https://issues.redhat.com//browse/OSDOCS-13342)
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview:
https://88848--ocpdocs-pr.netlify.app/openshift-enterprise/latest/extensions/catalogs/managing-catalogs.html
https://88848--ocpdocs-pr.netlify.app/openshift-enterprise/latest/extensions/ce/managing-ce.html
https://88848--ocpdocs-pr.netlify.app/openshift-enterprise/latest/operators/olm_v1/index.html
https://88848--ocpdocs-pr.netlify.app/openshift-enterprise/latest/extensions/arch/operator-controller#olmv1-clusterextension-api_operator-controller
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review:
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

michaelryanpeter commented on modules/olmv1-finding-operators-to-install.adoc:
```diff
@@ -24,10 +24,18 @@ include::snippets/olmv1-on-cluster-catalog-queries.adoc[]
-  and (.properties[] | select(.type == "olm.csv.metadata").value.installModes[] \
-  | select(.type == "AllNamespaces" and .supported == true)) \
-  and .spec.webhookdefinitions == null) | .package] | unique[]'
+  | jq -r --arg pkg "" ' select(.schema == "olm.bundle" and (.package == $pkg or $pkg == "")) \
+  | {package:.package, name:.name, image:.image, supportsAllNamespaces: (.properties[] \
+  | select(.type == "olm.csv.metadata").value.installModes[] \
+  | select(.type == "AllNamespaces").supported == true)} ' \
+  | tee allNamespaces.json | jq -r '.image' \
+  | xargs -I '{}' -n1 -P8 bash -c ' opm render {} > $(mktemp -d -p . -t olmv1-compat-bundle-XXXXXXXX)/bundle.json ' && bash -c 'cat olmv1-compat-bundle*/*.json' \
+  | jq '{package, name, image, requiresWebhooks: (.properties[] \
+  | select(.type == "olm.bundle.object").value.data | @base64d | fromjson \
+  | select(.kind == "ClusterServiceVersion").spec.webhookdefinitions != null)}' > webhooks.json && jq -cs ' group_by(.name)[] \
+  | reduce .[] as $item ({}; . *= $item) | . *= {compatible: ((.requiresWebhooks | not) and .supportsAllNamespaces)} \
+  | {name, package, compatible} ' allNamespaces.json webhooks.json | jq -r '. \
+  | select(.compatible == true) | .package' | sort -u
```
Happy to remove these queries and go back to the old ones. I've updated this as a separate commit, so it is easy to remove.

michaelryanpeter commented on modules/olmv1-finding-operators-to-install.adoc:
```diff
@@ -41,10 +49,18 @@ where:
-  and (.properties[] | select(.type == "olm.csv.metadata").value.installModes[] \
-  | select(.type == "AllNamespaces" and .supported == true)) \
-  and .spec.webhookdefinitions == null) | .package] | unique[]'
+  | jq -r --arg pkg "" ' select(.schema == "olm.bundle" and (.package == $pkg or $pkg == "")) \
+  | {package:.package, name:.name, image:.image, supportsAllNamespaces: (.properties[] \
+  | select(.type == "olm.csv.metadata").value.installModes[] \
+  | select(.type == "AllNamespaces").supported == true)} ' \
+  | tee allNamespaces.json | jq -r '.image' \
+  | xargs -I '{}' -n1 -P8 bash -c ' opm render {} > $(mktemp -d -p . -t olmv1-compat-bundle-XXXXXXXX)/bundle.json ' && bash -c 'cat olmv1-compat-bundle*/*.json' \
+  | jq '{package, name, image, requiresWebhooks: (.properties[] \
+  | select(.type == "olm.bundle.object").value.data | @base64d | fromjson \
+  | select(.kind == "ClusterServiceVersion").spec.webhookdefinitions != null)}' > webhooks.json && jq -cs ' group_by(.name)[] \
+  | reduce .[] as $item ({}; . *= $item) | . *= {compatible: ((.requiresWebhooks | not) and .supportsAllNamespaces)} \
+  | {name, package, compatible} ' allNamespaces.json webhooks.json | jq -r '. \
+  | select(.compatible == true) | .package' | sort -u
```
Happy to remove these queries and go back to the old ones. I've updated this as a separate commit, so it is easy to remove.

michaelryanpeter commented on modules/olmv1-finding-operators-to-install.adoc:
```diff
@@ -77,24 +93,39 @@ $ opm render \
 $ opm render <catalog_registry_url>:<tag> \
-  | jq -s '.[] | select( .schema == "olm.package") \
-  | select( .name == "<package_name>")'
+  | jq -r --arg pkg "openshift-pipelines-operator-rh" 'select(.schema == "olm.bundle" and (.package == $pkg or $pkg == "")) \
+  | {"package":.package, "name":.name, "image":.image, "supportsAllNamespaces": (.properties[] \
+  | select(.type == "olm.csv.metadata").value.installModes[] \
+  | select(.type == "AllNamespaces").supported == true)}' | tee allNamespaces.json | jq -r '.image' \
+  | xargs -I '{}' -n1 -P8 bash -c 'opm render {} > $(mktemp -d -p . -t olmv1-compat-bundle-XXXXXXXX)/bundle.json' && bash -c 'cat olmv1-compat-bundle*/*.json' \
+  | jq '{package, name, image, "requiresWebhooks": (.properties[] \
+  | select(.type == "olm.bundle.object").value.data | @base64d | fromjson \
+  | select(.kind == "ClusterServiceVersion").spec.webhookdefinitions != null)}' > webhooks.json && jq -cs 'group_by(.name)[] \
+  | reduce .[] as $item ({}; . *= $item) | . *= {"compatible": ((.requiresWebhooks | not) and .supportsAllNamespaces)} \
+  | {name,compatible}' allNamespaces.json webhooks.json && rm -rf allNamespaces.json webhooks.json olmv1-compat-bundle*
```
same comment as the other updated `jq` query

michaelryanpeter commented on modules/olmv1-finding-operators-to-install.adoc:
```diff
@@ -77,24 +93,39 @@ $ opm render \
-  registry.redhat.io/redhat/redhat-operator-index:v{product-version} \
-  | jq -s '.[] | select( .schema == "olm.package") \
-  | select( .name == "openshift-pipelines-operator-rh")'
+$ opm render registry.redhat.io/redhat/redhat-operator-index:v{product-version} \
+  | jq -r --arg pkg "openshift-pipelines-operator-rh" 'select(.schema == "olm.bundle" and (.package == $pkg or $pkg == "")) \
+  | {"package":.package, "name":.name, "image":.image, "supportsAllNamespaces": (.properties[] \
+  | select(.type == "olm.csv.metadata").value.installModes[] \
+  | select(.type == "AllNamespaces").supported == true)}' | tee allNamespaces.json | jq -r '.image' \
+  | xargs -I '{}' -n1 -P8 bash -c 'opm render {} > $(mktemp -d -p . -t olmv1-compat-bundle-XXXXXXXX)/bundle.json' && bash -c 'cat olmv1-compat-bundle*/*.json' \
+  | jq '{package, name, image, "requiresWebhooks": (.properties[] \
+  | select(.type == "olm.bundle.object").value.data | @base64d | fromjson \
+  | select(.kind == "ClusterServiceVersion").spec.webhookdefinitions != null)}' > webhooks.json && jq -cs 'group_by(.name)[] \
+  | reduce .[] as $item ({}; . *= $item) | . *= {"compatible": ((.requiresWebhooks | not) and .supportsAllNamespaces)} \
+  | {name,compatible}' allNamespaces.json webhooks.json && rm -rf allNamespaces.json webhooks.json olmv1-compat-bundle*
```
same comment as the other updated `jq` query

michaelryanpeter commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -16,20 +16,22 @@ To add a catalog to a cluster for {olmv1-first} usage, create a `ClusterCatalog`
-<1> Specify the catalog's image in the `spec.source.image` field.
-<2> Specify the interval for polling the remote registry for newer image digests. The default value is `24h`. Valid units include seconds (`s`), minutes (`m`), and hours (`h`). To disable polling, set a zero value, such as `0s`.
-
+<1> The catalog is automatically labelled with the value of the `metadata.name` field when it is applied to the cluster. For more information about labels and catalog selection, see "Catalog content resolution".
```
Addressed this typo locally. I am waiting to push the change because builds are taking so long.

michaelryanpeter commented:
@kuiwang02 or @jianzhangbjz Would you PTAL?

kuiwang02 commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -16,19 +16,22 @@ To add a catalog to a cluster for {olmv1-first} usage, create a `ClusterCatalog`
     image:
-      ref: registry.redhat.io/redhat/redhat-operator-index:v{product-version} <1>
-      pollInterval: <poll_interval_duration> <2>
+      pollIntervalMinutes: 10 <3>
```
@michaelryanpeter pollIntervalMinutes is different with the previous release.
if we do not want to poll it, do not set it. and there is no default value.

FYI https://github.com/operator-framework/operator-controller/blob/dd4d61fffa399470cf49b5506cf54c0175b40fd3/catalogd/api/v1/clustercatalog_types.go#L345-L351

michaelryanpeter replied:
Addressed: https://github.com/openshift/openshift-docs/pull/88848/files#diff-37c51032cc762a74feb1f038c6e924252809d045fd71909be26377b6fc82b4edR33

kuiwang02 commented on modules/olmv1-installing-an-operator.adoc:
```diff
@@ -29,15 +29,15 @@ spec:
   serviceAccount:
     name: <service_account>
-  channel: <channel>
+  channels: [<channel>]
```
@michaelryanpeter the fields of the example is not correct.
here is the current example.

```yaml
 apiVersion: olm.operatorframework.io/v1
  kind: ClusterExtension
  metadata:
    name: enough-75493
  spec:
    namespace: ns-75493
    serviceAccount:
      name: sa75493
    source:
      sourceType: Catalog
      catalog:
        packageName: nginx-ok-v3277775493
        channels:
          - alpha
        version: ">=0.0.1"
```

it seems you forget to update it because you have https://github.com/openshift/openshift-docs/pull/88848/files#diff-57d7e8d28879735eab92932f9cc5f093124acaf00427fc24c2ce84e78e186d99R87-R95

michaelryanpeter replied:
Addressed: https://github.com/openshift/openshift-docs/pull/88848/files#diff-57d7e8d28879735eab92932f9cc5f093124acaf00427fc24c2ce84e78e186d99R23-R37

kuiwang02 commented on modules/olmv1-installing-an-operator.adoc:
```diff
@@ -50,7 +50,7 @@ $ oc apply -f pipeline-operator.yaml
 [source,text]
 ----
-clusterextension.olm.operatorframework.io/pipelines-operator created
+clusterextension.olm.operatorframework.io/pipes created
```
@michaelryanpeter it is clusterextension.olm.operatorframework.io/pipelines-operator because of https://github.com/openshift/openshift-docs/pull/88848/files#diff-57d7e8d28879735eab92932f9cc5f093124acaf00427fc24c2ce84e78e186d99R26

michaelryanpeter replied:
Addressed: https://github.com/openshift/openshift-docs/pull/88848/files#diff-57d7e8d28879735eab92932f9cc5f093124acaf00427fc24c2ce84e78e186d99R57

kuiwang02 commented on modules/olmv1-installing-an-operator.adoc:
```diff
@@ -69,76 +69,76 @@ $ oc get clusterextension pipelines-operator -o yaml
-    name: pipelines-operator
-    resourceVersion: "53324"
-    uid: c54237be-cde4-46d4-9b31-d0ec6acc19bf
+    name: pipes
```
@michaelryanpeter same above comment.

michaelryanpeter replied:
Addressed: https://github.com/openshift/openshift-docs/pull/88848/files#diff-57d7e8d28879735eab92932f9cc5f093124acaf00427fc24c2ce84e78e186d99R87

kuiwang02 commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -187,9 +194,9 @@ metadata:
   packageName: openshift-pipelines-operator-rh
   installNamespace: <namespace>
-  version: "1.12.1" <1>
+  version: "1.15.0" <1>
```
@michaelryanpeter the example is not correct. it still use old format, for example 
olm.operatorframework.io/v1alpha1
```yaml
spec:
  packageName: openshift-pipelines-operator-rh
  installNamespace: <namespace>
```

kuiwang02 commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -203,9 +210,9 @@ metadata:
   packageName: openshift-pipelines-operator-rh
   installNamespace: <namespace>
-  version: ">1.11.1, <1.13" <1>
+  version: ">1.15, <1.17" <1>
```
@michaelryanpeter same above

michaelryanpeter replied:
Addressed: https://github.com/openshift/openshift-docs/pull/88848/files#diff-3a4664f5b5ccef6dd8762bcc6bc754af9267fa277045a6f9aeabb7807fb32087L199-R223

kuiwang02 commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -219,7 +226,7 @@ metadata:
   packageName: openshift-pipelines-operator-rh
   installNamespace: <namespace>
-  channel: pipelines-1.13 <1>
+  channel: lastest <1>
```
@michaelryanpeter same above

kuiwang02 commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -236,7 +243,7 @@ spec:
   installNamespace: <namespace>
   channel: latest
-  version: "<1.13"
+  version: "<1.16"
```
@michaelryanpeter same above

michaelryanpeter replied:
Addressed: https://github.com/openshift/openshift-docs/pull/88848/files#diff-3a4664f5b5ccef6dd8762bcc6bc754af9267fa277045a6f9aeabb7807fb32087R254-R268

kuiwang02 commented on modules/olmv1-updating-an-operator.adoc:
```diff
@@ -251,26 +258,8 @@ $ oc apply -f pipelines-operator.yaml
 [source,text]
 ----
-clusterextension.olm.operatorframework.io/pipelines-operator configured
+clusterextension.olm.operatorframework.io/pipes configured
```
@michaelryanpeter the name pipes is not correct with your yaml.

michaelryanpeter replied:
Addressed: https://github.com/openshift/openshift-docs/pull/88848/files#diff-3a4664f5b5ccef6dd8762bcc6bc754af9267fa277045a6f9aeabb7807fb32087R283

jianzhangbjz commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -16,19 +16,22 @@ To add a catalog to a cluster for {olmv1-first} usage, create a `ClusterCatalog`
 [source,yaml,subs="attributes+"]
 ----
-apiVersion: catalogd.operatorframework.io/v1alpha1
+apiVersion: olm.operatorframework.io/v1
```
The same issue for https://docs.openshift.com/container-platform/4.18/extensions/arch/operator-controller.html

michaelryanpeter replied:
@kuiwang02 For the _Managing catalogs_ book, I updated the module on the Red Hat Operators in this PR: https://github.com/openshift/openshift-docs/pull/88394/files#diff-231ae6128db7e4335d2cb5e97d55a446d742c4f6159bb691969fbce29cec5ea9

I will make the rest of the corrections in this PR.

michaelryanpeter replied:
Operator controller API addressed here: https://github.com/openshift/openshift-docs/pull/88848/files#diff-ccc57dbd3b19c42337c5ef6c845c601be66555465127ca65aba2ced3474c32da

kuiwang02 replied:
ok to take care that in the PR #88349. thanks

michaelryanpeter commented:
@kuiwang02 Thank you for your review! I applied your feedback. PTAL.

michaelryanpeter commented:
/label peer-review-needed

kuiwang02 commented on modules/olmv1-installing-an-operator.adoc:
```diff
@@ -20,25 +20,29 @@ You can install an extension from a catalog by creating a custom resource (CR) a
 [source,yaml]
 ----
-apiVersion: olm.operatorframework.io/v1alpha1
+apiVersion: olm.operatorframework.io/v1
```
@michaelryanpeter when I review https://88848--ocpdocs-pr.netlify.app/openshift-enterprise/latest/extensions/ce/managing-ce.html, I find 
```yaml
opm render <catalog_registry_url>:<tag> | jq -s '.[] | select( .schema == "olm.package")
```
is not correct. it miss "'"

the correct is 
```yaml
opm render <catalog_registry_url>:<tag> | jq -s '.[] | select( .schema == "olm.package")'
```

I know it has nothing with this PR, but I do not want to create the bug which is late to fix, so I raise it here.

thanks to understand it.

michaelryanpeter replied:
I agree. Let's fix it while we're here!

kuiwang02 commented on modules/olmv1-installing-an-operator.adoc:
```diff
@@ -20,25 +20,29 @@ You can install an extension from a catalog by creating a custom resource (CR) a
+    sourceType: Catalog
+    catalog:
+      packageName: openshift-pipelines-operator-rh
+      channels:
```
@michaelryanpeter you list channels here, but there is no channels for your example.
I am not sure if it is ok from writing's view. if it is ok, please keep it.  


by the way, from my view,  it is better to provide the example with parameter, like
```yaml
apiVersion: olm.operatorframework.io/v1
  kind: ClusterExtension
  metadata:
    name: <clusterextension-name>
  spec:
    namespace: <installed-namespace>
    serviceAccount:
      name: <installed-sa-name>
    source:
      sourceType: Catalog
      catalog:
        packageName: <package-name>
        channels:
          - <channel-name>
        version: <versions>
        upgradeConstraintPolicy: <constraint-policy>
```
with the indication: channels, version, upgradeConstraintPolicy are optional.

and then provide the instance
```yaml
apiVersion: olm.operatorframework.io/v1
  kind: ClusterExtension
  metadata:
    name: pipelines-operator
  spec:
    namespace: pipelines
    serviceAccount:
      name: pipelines-installer
    source:
      sourceType: Catalog
      catalog:
        packageName: openshift-pipelines-operator-rh
        version: "1.14.x"
```
and it matches the output of your example.

it is better for user to understand it.

Thanks

michaelryanpeter replied:
Great feedback. Thank you!

michaelryanpeter replied:
I am going to take your comment, but I am going to leave off the `upgradeConstraintPolicy` field for the moment. I think I should better describe the risks associated with choosing `SelfCertified`. I created a ticket to follow up on this work next sprint: https://issues.redhat.com/browse/OSDOCS-13418

kuiwang02 replied:
ok, thanks

adellape commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -16,19 +16,22 @@ To add a catalog to a cluster for {olmv1-first} usage, create a `ClusterCatalog`
-      ref: registry.redhat.io/redhat/redhat-operator-index:v{product-version} <1>
-      pollInterval: <poll_interval_duration> <2>
+      pollIntervalMinutes: 10 <3>
+      ref: registry.redhat.io/redhat/community-operator-index:vBranch Build <4>
```
```suggestion
      ref: registry.redhat.io/redhat/community-operator-index:v{product-version} <4>
```
?

michaelryanpeter replied:
whoops. good catch.

michaelryanpeter commented:
Thank you for the review, @kuiwang02. I addressed your comments in this commit: https://github.com/openshift/openshift-docs/pull/88848/commits/ec4d00bf7653127e6b62c0517a90639cc9a902a2

I do think I should wait to document the `upgradeConstraintPolicy` field until we can make sure customers understand it and don't accidentally have a problem.

Would you PTAL?

adellape commented on modules/olmv1-adding-a-catalog.adoc:
```diff
@@ -16,19 +16,22 @@ To add a catalog to a cluster for {olmv1-first} usage, create a `ClusterCatalog`
+<1> The catalog is automatically labeled with the value of the `metadata.name` field when it is applied to the cluster. For more information about labels and catalog selection, see "Catalog content resolution".
+<2> Optional: Specify the priority of the catalog in relation to the other catalogs on the cluster. For more information, see "Catalog selection by priority".
+<3> Specify the interval in minutes for polling the remote registry for newer image digests. To disable polling, do not set the field.
+<4> Specify the catalog's image in the `spec.source.image.ref` field.
```
Same?
```suggestion
<4> Specify the catalog image in the `spec.source.image.ref` field.
```

kuiwang02 commented:
> Thank you for the review, @kuiwang02. I addressed your comments in this commit: [ec4d00b](https://github.com/openshift/openshift-docs/commit/ec4d00bf7653127e6b62c0517a90639cc9a902a2)
> 
> I think I should wait to document the `upgradeConstraintPolicy` field until we can write more docs to make sure customers understand it and don't accidentally have a problem. WDYT?
> 
> Would you PTAL?

it is ok for me. thanks

/lgtm

michaelryanpeter commented:
/cherrypick enterprise-4.18
